util/tracecpu/profile.go

```diff
@@ -0,0 +1,432 @@
+// The `sqlStats` maybe:
+//     plans: {
+//         "table_scan": 200ms, // The cpu time of the sql that plan with `table_scan` is 200ms.
+//         "index_scan": 300ms, // The cpu time of the sql that plan with `table_scan` is 300ms.
+//       },
+//     total:      600ms,       // The total cpu time of the sql is 600ms.
+// total_time - table_scan_time - index_scan_time = 100ms, and this 100ms means those sample data only contain the
+// sql_digest label, doesn't contain the plan_digest label. This is cause by the `pprof profile` is base on sample.
+// After this tune function, the `sqlStats` become to:
+//     plans: {
+//         "table_scan": 240ms, // 200 + (200/(200+300))*100
+//         "index_scan": 360ms, // 300 + (300/(200+300))*100
+//       },
+//     total:      600ms,
+func (s *sqlStats) tune() {
+	if len(s.plans) == 0 {
+		s.plans[""] = s.total
+		return
+	}
+	planTotal := int64(0)
+	for _, v := range s.plans {
+		planTotal += v
+	}
+	remain := s.total - planTotal
+	if remain <= 0 {
+		return
+	}
+	for k, v := range s.plans {
+		s.plans[k] = v + (v/planTotal)*remain
+	}
+}
+
+func (sp *sqlStatsProfiler) handleExportProfileTask(p *profile.Profile) {
+	sp.mu.Lock()
+	defer sp.mu.Unlock()
+	if sp.mu.ept == nil {
+		return
+	}
+	sp.mu.ept.mergeProfile(p)
+}
+
+func (sp *sqlStatsProfiler) hasExportProfileTask() bool {
+	sp.mu.Lock()
+	has := sp.mu.ept != nil
+	sp.mu.Unlock()
+	return has
+}
+
+// IsEnabled return true if it is(should be) enabled. It exports for tests.
+func (sp *sqlStatsProfiler) IsEnabled() bool {
+	return config.GetGlobalConfig().TopSQL.Enable || sp.hasExportProfileTask()
+}
+
+// StartCPUProfile same like pprof.StartCPUProfile.
+// Because the GlobalSQLStatsProfiler keep calling pprof.StartCPUProfile to fetch SQL cpu stats, other place (such pprof profile HTTP API handler) call pprof.StartCPUProfile will be failed,
+// other place should call tracecpu.StartCPUProfile instead of pprof.StartCPUProfile.
+func StartCPUProfile(w io.Writer) error {
+	if GlobalSQLStatsProfiler.IsEnabled() {
+		return GlobalSQLStatsProfiler.startExportCPUProfile(w)
+	}
+	return pprof.StartCPUProfile(w)
+}
+
+// StopCPUProfile same like pprof.StopCPUProfile.
+// other place should call tracecpu.StopCPUProfile instead of pprof.StopCPUProfile.
+func StopCPUProfile() error {
+	if GlobalSQLStatsProfiler.IsEnabled() {
+		return GlobalSQLStatsProfiler.stopExportCPUProfile()
+	}
+	pprof.StopCPUProfile()
+	return nil
+}
+
+// SetGoroutineLabelsWithSQL sets the SQL digest label into the goroutine.
+func SetGoroutineLabelsWithSQL(ctx context.Context, normalizedSQL, sqlDigest string) context.Context {
+	if len(normalizedSQL) == 0 || len(sqlDigest) == 0 {
+		return ctx
+	}
+	if variable.EnablePProfSQLCPU.Load() {
+		ctx = pprof.WithLabels(context.Background(), pprof.Labels(labelSQLDigest, sqlDigest, labelSQL, util.QueryStrForLog(normalizedSQL)))
+	} else {
+		ctx = pprof.WithLabels(context.Background(), pprof.Labels(labelSQLDigest, sqlDigest))
```

Comment:
maybe we should `context.Backgroud` -> `ctx`?

Reply:
Oh, nice catch, it a bug here. 
done.